designs/rerun_cli_commands.md

```diff
@@ -0,0 +1,85 @@
+# Rerun CLI commands after failure
+
+## Introduction
+**Problem:**
+Most of the errors returned by the CLI cluster commands are either transient or fixable with manual intervention.
+However, the CLI commands are not idempotent.
+This makes impossible to rerun commands after they have already failed, even if the root cause of the issues has already been solved.
+This makes the experience very disruptive, specially for the upgrade cluster command, which could leave the cluster in an irrecoverable state without the necessary manual cleanup.
+This "cleanup" is not documented and it heavily dependents on the internal CLI's implementation.
+For certain scenarios, it is not even possible, requiring users to destroy and recreate their clusters from scratch.
+
+## Solution
+The proposed solution is to implement a "checkpoint" capability.
+This will serve as the first step to a larger story of improving the customer experience with upgrading clusters and troubleshooting in general.
+The CLI will keep a registry of all completed steps for a workflow and the necessary data to restore its state.
+When a command is rerun after an error, the CLI will skip the completed steps, restoring the state of the program.
+The checkpoint data would be stored in the `<clusterName>/generated` folder as a file named `<clusterName>-checkpoint.yaml`.
+
+Example checkpoint file:
+
+```yaml
+completedTasks:
+  bootstrap-cluster-init:
+    ExistingManagement: false
+    KubeconfigFile: test/generated/test.kind.kubeconfig
+    Name: test
+  ensure-etcd-capi-components-exist: null
+  install-capi: null
+  pause-controllers-reconcile: null
+  setup-and-validate: null
+  update-secrets: null
+  upgrade-core-components:
+    components: []
+  upgrade-needed: null
+```
+
+When the task finishes with no errors, we will add it as a completed task in the checkpoint file with the necessary information (if any) to allow the command to rerun successfully.
```

Comment:
I wonder if we should add more validations to each step before updating the checkpoint step to be "true/success". or maybe validate each succeeded step when we start the rerun